nodejs/week2/app.js

```diff
@@ -0,0 +1,71 @@
+const express = require("express");
+const app = express();
+const port = 2000;
+const users = require("./documents.json");
+app.use(express.json());
+
+// GET /search
+app.get("/search", (req, res) => {
+  let value = req.query.q;
+  if (value) {
+    const result = filterByQuery(value);
+    res.send(result);
+  } else {
+    return res.send(users);
+  }
+});
+
+app.get("/documents/:id", (req, res) => {
+  const id = parseInt(req.params.id);
+  const response = users.find((obj) => obj.id === id);
+  if (!response) {
+    res.status(404).send("user not found");
+  }
+  res.json(response);
+});
+
+app.post("/search", (req, res) => {
+  // check if req.body = fields;
+  let fields = req.body.fields;
+  // check if req has q.param
+  let q = req.query.q;
+  // if q and fields not embty (bad request)
+  if (fields && q) return res.status(400);
+  // if q search
+  if (q) {
+    const result = filterByQuery(q);
+    return res.send(result);
+  }
+  // else search for fields with value
+  if (fields) {
+    const result = filterByFields(fields);
+    return res.send(result);
+  }
+  // if q and fields are empty
+  return res.status(400);
+});
+
+app.listen(port, () => {
+  console.log(`Listening on port ${port}`);
+});
+
+function filterByQuery(value) {
+  return users.filter((user) => {
+    return arrayOfAttributesOfUser.some(([key, val]) =>
+      String(val).toLowerCase().includes(value.toLowerCase())
+    );
+  });
+}
```

Comment:
Did you mean to use `user` inside the filter function?
`arrayOfAttributesOfUser` is never defined making the search  by query fail. 
Array.some probably does not work in this case as `user` is an object. The filter function will look at the objects in the array one by one.
I would suggest looking at [Object.entries()](https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Global_Objects/Object/entries) or [Object.values()](https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Global_Objects/Object/values)

Let me know if anything needs more explanation!